### ISSUE TITLE
Link definitions not propogated to nested parser

### ISSUE THREAD

**Issue by chrisjsewell** [ExecutableBookProject/MyST-Parser#113]
Currently, link definitions cannot be used inside directives, e.g.

````
```{note}
[reference]
```

[reference]: https://github.com/ExecutableBookProject
````

will not work, but this will:

````
```{note}
[reference]

[reference]: https://github.com/ExecutableBookProject
```

````

This will also be the case for upcoming footnotes

**Comment by chrisjsewell** [comment]
there are two places where a new document is parsed:

- https://github.com/ExecutableBookProject/MyST-Parser/blob/c64f7068ba786a79561adb0017194ebc8bf15ba3/myst_parser/docutils_renderer.py#L108
- https://github.com/ExecutableBookProject/MyST-Parser/blob/c64f7068ba786a79561adb0017194ebc8bf15ba3/myst_parser/docutils_renderer.py#L838

The first fix, is that both should be called with `reset_definitions=False`, to ensure we do not wipe currently collected definitions from the global parse context.

For context, these references are only resolved if the definition is immediately available. To achieve this for a normal document parse in mistletoe, we run a first parse of the source text, where span tokens are not yet processed (stored as raw strings in `SpanContainer`), then a subsequent 'walk' of the syntax tree to process these containers.

**Comment by chrisjsewell** [comment]
@choldgraf something to consider in MyST-NB, currently during the `Document.read` (per cell) definitions are reset; so definitions in one text cell cannot be used by a reference in another cell. This is a different behaviour to if the notebook was e.g. converted to a text document with Jupytext and parsed as a whole.

https://github.com/ExecutableBookProject/MyST-NB/blob/c9478b70daeb34be30c48e807e3aa2ec5611f420/myst_nb/parser.py#L78

**Comment by choldgraf** [comment]
Interesting - how do you imagine it would be better to convert the `ipynb` to a text file (maybe with `{execute}` markdown cells?) and parsing that directly? One downside there is that then `ipynb` *with outputs* wouldn't work.

Is there a way that we could persist information like footnotes etc across cell tokenizing?

**Comment by chrisjsewell** [comment]
I'm not sure of an ideal solution off-hand; just somehow you need to 'defer' the processing of Markdown span tokens until you've found all possible link/footnote definitions are found.

It's a bit of a peculiarity of Markdown, that for example, you can't just identify the references through a regex, then store them for later replacement with the definitions, because whether or not they can be immediately resolved has a direct bearing on how the subsequent text is parsed.

For example, this:

```
[a*b]*

[a*b]: asd
```

Is parsed to this

[a*b]*

[a*b]: asd

but if the definition is not available:

```
[c*d]*
```

[c*d]*

**Comment by chrisjsewell** [comment]
Actually I'm not sure if this is strictly part of the CommonMark spec (see [this interactive demo](https://spec.commonmark.org/dingus/?text=%5Ba*b%5D*c%0A%0A%5Ba*b%5D%3A%20asd%0A%0A%5Bc*d%5D*e)). I'll have to see in mistletoe-ebp, if storing all potential definition reference as 'pending references' breaks any tests, because this would make things a little easier.

**Comment by chrisjsewell** [comment]
> Actually I'm not sure if this is strictly part of the CommonMark spec (see [this interactive demo](https://spec.commonmark.org/dingus/?text=%5Ba*b%5D*c%0A%0A%5Ba*b%5D%3A%20asd%0A%0A%5Bc*d%5D*e)). I'll have to see in mistletoe-ebp, if storing all potential definition reference as 'pending references' breaks any tests, because this would make things a little easier.

Yeh definitely not possible, or at least very difficult to capture the following kind of parsing, whereby the presence of the definition massively changes the parsing flow: 

```
[link [foo [bar]]](/uri)

[bar]: a

[link [foo [other]]](/uri)
```

[link [foo [bar]]](/uri)

[bar]: a

[link [foo [other]]](/uri)

**Comment by choldgraf** [comment]
so does this mean that for now we just need to tell people "if you want to use a link replacement with `ipynb` files in jupyter book, you need to have the link definitions in the same cell"?

**Comment by mmcky** [comment]
@choldgraf @chrisjsewell this is a pretty important issue I think. My feeling is the `ipynb` document needs to be wholistic rather than tokenistic (based on cells).  Can we pass the AST between reads of the `cells`? 

Given the duality between `ipynb` and `myst` means the entry point can be from either format so we should also consider `myst` text as the entry point.

**Comment by chrisjsewell** [comment]
> @choldgraf @chrisjsewell this is a pretty important issue I think. My feeling is the `ipynb` document needs to be wholistic rather than tokenistic (based on cells). Can we pass the AST between reads of the `cells`?

This is my feeling;  the thing is we need to build the **full Markdown AST**, before applying the renderer (i.e. Markdown AST -> docutils AST). This is mainly possible, with some restructuring of the MyST-NB parser but, as I've said before, precludes this notion of wrapping text cells in coutils elements, like here: https://github.com/ExecutableBookProject/MyST-NB/blob/c9478b70daeb34be30c48e807e3aa2ec5611f420/myst_nb/parser.py#L80

**Comment by choldgraf** [comment]
Shall we take this issue over to myst-nb then?

**Comment by chrisjsewell** [comment]
Maybe make a separate issue on. In the first instance the is a myst-parser bug

**Comment by chrisjsewell** [comment]
closed in #119
